### ISSUE TITLE
New Rule: Enforce De Morgan transformations in boolean expressions

### ISSUE THREAD

**Issue by azat-io** [eslint/eslint#19400]
### Rule details

Add rule to enforce De Morgan transformations in boolean expressions.

### Related ECMAScript feature

None. De Morgan’s laws are a classical logical transformation that has been part of ECMAScript since its inception.

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

It would be great to have an ESLint rule that enforces (or optionally enforces) the application of [De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) in boolean expressions. For example, the rule should flag and offer an autofix for code such as:

```js
if (!(a && b)) { /* ... */ }
```

and suggest transforming it into:

``` js
if (!a || !b) { /* ... */ }
```

Similarly, for negated disjunctions:

```js
if (!(a || b)) { /* ... */ }
```

the rule would enforce rewriting it as:

```js
if (!a && !b) { /* ... */ }
```

### Why should this rule be in the core instead of a plugin?

This increases the readability of the code. And it would be a good tool for refactoring.

Many coding style guides and developers prefer the explicit form using De Morgan's law because it can improve readability and consistency across a codebase. 

An optional rule to enforce this pattern (with an auto-fix capability) could help maintain a consistent coding style in projects that value clear boolean logic.

The rule should detect expressions of the form `!(A && B)` and `!(A || B)`.

Suggestion/Autofix:
- For `!(A && B)`, suggest rewriting as `!A || !B`.  
- For `!(A || B)`, suggest rewriting as `!A && !B`.

### Participation

- [x] I am willing to submit a pull request to implement this rule.

### Additional comments

N/A.

**Comment by mdjermanovic** [comment]
Hi @azat-io, thanks for the rule proposal. 

I think this rule doesn't qualify for inclusion in the core, per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies/), as it's a stylistic rule that isn't related to new ECMAScript features.

@eslint/eslint-team any other thoughts?

**Comment by nzakas** [comment]
I agree. We aren't including rules that "improve readability," as this is highly subjective.

You can still create a custom rule that does this, though.

**Comment by azat-io** [comment]
Thank you for your response!

I've created a ESLint plugin for this:
https://github.com/azat-io/eslint-plugin-de-morgan
